### PR TITLE
Allow add config from django model

### PR DESCRIPTION
**Pull request by luismorenolopera**

### CODE REVIEW

File: seb_openedx/models.py [hunk 1]
```diff
@@ -23,3 +32,44 @@ class Meta(object):
         indexes = [
             models.Index(fields=['username', 'course_id'])
         ]
+
+
+class SebCourseConfiguration(models.Model):
+    """Model that stores settings per course for seb browser."""
+    course_id = CourseKeyField(max_length=255, unique=True)
+    permission_components = models.TextField(blank=True, default='AlwaysAllowStaff.CheckSEBHashBrowserExamKey.CheckSEBHashConfigKey')
```
Comment (felipemontoya) [inline seb_openedx/models.py]: Please use , instead (or \n)

Comment (felipemontoya) [reply seb_openedx/models.py]: We could get the orginal defaults and convert them to unicode here

File: seb_openedx/models.py [hunk 2]
```diff
@@ -23,3 +32,44 @@ class Meta(object):
...
+    def __unicode__(self):
+        """Formats impression of object."""
+        return six.text_type(self.course_id)
+
+    @classmethod
+    def get_config_by_course_id(cls, course_id):
+        """Get config by course_id."""
+        queryset = cls.objects.get(course_id=course_id)
+        return {
+            'PERMISSION_COMPONENTS': queryset.permission_components.split('.'),
```
Comment (felipemontoya) [inline seb_openedx/models.py]: the split character should be a constant

File: seb_openedx/widgets.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        - 'val1.val2.val3'.
+    """
+    template_name = 'widgets/list_widget.html'
+
+    def get_context(self, name, value, attrs):
+        """
+        Format the values ​​before sending them to the template.
+        """
+        try:
+            value = value.split('.')
```
Comment (felipemontoya) [inline seb_openedx/widgets.py]: Use the constant

File: seb_openedx/widgets.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+"""
+Widgets for seb-openedx froms.
```
Comment (felipemontoya) [inline seb_openedx/widgets.py]: typo

File: seb_openedx/widgets.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+"""
+from django.forms.widgets import Widget
+
+
+class ListWidget(Widget):
+    """
+    Shows in the template with the break line format for text with next format:
+        - 'val1.val2.val3'.
```
Comment (felipemontoya) [inline seb_openedx/widgets.py]: You could explain a bit more

File: seb_openedx/admin.py [hunk 6]
```diff
@@ -1,7 +1,122 @@
...
+
+from seb_openedx.models import (
+    ForbiddenCourseAccess,
+    SebCourseConfiguration,
+    SebCourseConfigurationHistory
+)
+from seb_openedx.widgets import ListWidget
+
+
+class SebCourseConfigurationForm(forms.ModelForm):
```
Comment (felipemontoya) [inline seb_openedx/admin.py]: This should go to a `forms.py` file, right?

File: seb_openedx/admin.py [hunk 7]
```diff
@@ -1,7 +1,122 @@
...
+                                    courseware <br>
+                                    discussion <br>
+                                    progress <br>
+                                    wiki""")
+        }
+
+    def _format_array_field(self, data_field):
+        """Adapt array field content from breakline format to 'val1.val2.val3'."""
+        result = data_field.split('\r\n')
+        result = '.'.join(result)
```
Comment (felipemontoya) [inline seb_openedx/admin.py]: use the constant

File: seb_openedx/admin.py [hunk 8]
```diff
@@ -1,7 +1,122 @@
...
+    list_display = [
+        'course_id',
+        'user_banning_enabled',
+    ]
+    search_fields = ('course_id', )
+    formfield_overrides = {
+        TextField: {'widget': ListWidget}
+    }
+    form = SebCourseConfigurationForm
+    # object_history_template = 'seb-history.html'
```
Comment (felipemontoya) [inline seb_openedx/admin.py]: If it is not used, then remove. Lets no leave unsed commented code

File: seb_openedx/seb_keys_sources.py [hunk 9]
```diff
@@ -6,12 +6,22 @@
...
 
 LOG = logging.getLogger(__name__)
 
 
+def from_django_model(course_key):
+    """Get SEB keys from model SEBCourseSettings."""
+    try:
+        model_settings = SebCourseConfiguration.get_config_by_course_id(course_key)
+    except SebCourseConfiguration.DoesNotExist:
+        model_settings = {}
```
Comment (felipemontoya) [inline seb_openedx/seb_keys_sources.py]: revisar si es un None

Comment (felipemontoya) [reply seb_openedx/seb_keys_sources.py]: If the ProgrammingError ocurrs, log and cacht it.

File: seb_openedx/models.py [hunk 10]
```diff
@@ -23,3 +32,44 @@ class Meta(object):
...
+    class Meta(object):
+        """Meta."""
+        app_label = 'seb_openedx'
+
+    def __unicode__(self):
+        """Formats impression of object."""
+        return six.text_type(self.course_id)
+
+    @classmethod
+    def get_config_by_course_id(cls, course_id):
```
Comment (felipemontoya) [inline seb_openedx/models.py]: get_compatible_config_by_course_id

**Review by felipemontoya — CHANGES_REQUESTED** [review]
Good work.
Some inline comments and we should also be splitting this PR into two

File: seb_openedx/admin.py [hunk 11]
```diff
@@ -1,7 +1,27 @@
...
+    SebCourseConfiguration,
+)
+from seb_openedx.widgets import ListWidget
+
+
+class SebCourseConfigurationAdmin(admin.ModelAdmin):
+    """Admin for the SebCourseConfiguration model."""
+    list_display = [
+        'course_id',
+        'user_banning_enabled',
```
Comment (felipemontoya) [inline seb_openedx/admin.py:22]: I would like to see something about the keys in the list_display view.
It does not have to be part of this PR, but it has to be included before this is final

File: seb_openedx/admin.py [hunk 12]
```diff
@@ -1,7 +1,27 @@
...
+    ]
+    search_fields = ('course_id', )
```
Comment (felipemontoya) [inline seb_openedx/admin.py:29]: we should also search by the keys

File: seb_openedx/constants.py [hunk 13]
```diff
@@ -0,0 +1,4 @@
+"""
+Constants for seb-openedx.
+"""
+COMMA = ','
```
Comment (felipemontoya) [inline seb_openedx/constants.py]: Let's call this SEPARATOR_CHAR

File: seb_openedx/models.py [hunk 14]
```diff
@@ -23,3 +25,35 @@ class Meta(object):
         indexes = [
             models.Index(fields=['username', 'course_id'])
         ]
+
+
+class SebCourseConfiguration(models.Model):
+    """Model that stores settings per course for seb browser."""
+    course_id = CourseKeyField(max_length=255, unique=True)
+    permission_components = models.TextField(blank=True, default='AlwaysAllowStaff,CheckSEBHashBrowserExamKey,CheckSEBHashConfigKey')
```
Comment (felipemontoya) [inline seb_openedx/models.py]: Better defaults

Comment (felipemontoya) [reply seb_openedx/models.py]: something like `SEPARATOR_CHAR.joint(settings.DEFAULT_...SOURCES)`

File: seb_openedx/models.py [hunk 15]
```diff
@@ -23,3 +25,35 @@ class Meta(object):
...
+    browser_keys = models.TextField(blank=True)
+    config_keys = models.TextField(blank=True)
+    user_banning_enabled = models.BooleanField(default=False)
+    blacklist_chapters = models.TextField(blank=True)
+    whitelist_paths = models.TextField(blank=True, default='wiki,about')
```
Comment (felipemontoya) [inline seb_openedx/models.py]: Better defaults. `SEPARATOR_CHAR.join(settings.WHATEVS)`

File: seb_openedx/models.py [hunk 16]
```diff
@@ -23,3 +25,35 @@ class Meta(object):
...
         ]
+
+
+class SebCourseConfiguration(models.Model):
+    """Model that stores settings per course for seb browser."""
+    course_id = CourseKeyField(max_length=255, unique=True)
+    permission_components = models.TextField(blank=True, default='AlwaysAllowStaff,CheckSEBHashBrowserExamKey,CheckSEBHashConfigKey')
+    browser_keys = models.TextField(blank=True)
+    config_keys = models.TextField(blank=True)
+    user_banning_enabled = models.BooleanField(default=False)
```
Comment (felipemontoya) [inline seb_openedx/models.py]: better default

File: seb_openedx/models.py [hunk 17]
```diff
@@ -23,3 +25,35 @@ class Meta(object):
...
+        app_label = 'seb_openedx'
+
+    def __unicode__(self):
+        """Formats impression of object."""
+        return six.text_type(self.course_id)
+
+    @classmethod
+    def get_config_by_course_id(cls, course_id):
+        """Get config by course_id."""
+        queryset = cls.objects.get(course_id=course_id)
```
Comment (felipemontoya) [inline seb_openedx/models.py]: Having the name queryset for this object is a little confusing. I suggest to use instance.

File: seb_openedx/seb_keys_sources.py [hunk 18]
```diff
@@ -6,12 +6,25 @@
...
 
 
+def from_django_model(course_key):
+    """Get SEB settings from model SebCourseConfiguration."""
+    try:
+        model_settings = SebCourseConfiguration.get_config_by_course_id(course_key)
+    except SebCourseConfiguration.DoesNotExist:
+        model_settings = None
+    except ProgrammingError:
+        LOG.warning('Verify that the migrations were successfully executed')
```
Comment (felipemontoya) [inline seb_openedx/seb_keys_sources.py]:
```suggestion
        LOG.warning('Verify that the migrations for the `seb-openedx` app were successfully executed')
```

File: seb_openedx/templates/seb-history.html [hunk 19]
```diff
@@ -0,0 +1,42 @@
+{% extends "admin/base_site.html" %}
```
Comment (felipemontoya) [inline seb_openedx/templates/seb-history.html]: This template should not exist in the PR

**Review by felipemontoya — CHANGES_REQUESTED** [review]

**Comment by felipemontoya** [comment]
We are still missing the migration file and removing the parts that go to a different PR

File: seb_openedx/seb_keys_sources.py [hunk 20]
```diff
@@ -6,12 +6,27 @@
...
 
 
+def from_django_model(course_key):
+    """Get SEB settings from model SebCourseConfiguration."""
+    try:
+        model_settings = SebCourseConfiguration.get_compatible_config_by_course_id(course_key)
+    except SebCourseConfiguration.DoesNotExist:
+        model_settings = None
+    except ProgrammingError:
+        message = ("SebCourseConfiguration table not found, "
```
Comment (felipemontoya) [inline seb_openedx/seb_keys_sources.py:25]: py2?

File: seb_openedx/admin.py (hunk 12 above)
Comment (luismorenolopera) [reply seb_openedx/admin.py:29]: Done in method get_search_results in https://github.com/eduNEXT/seb-openedx/pull/39/files#diff-d5de91ba936faae76d26f21daa0372c7R35

File: seb_openedx/seb_keys_sources.py (hunk 20 above)
Comment (luismorenolopera) [reply seb_openedx/seb_keys_sources.py:25]: yes

**Comment by felipemontoya** [comment]
I tested on the ironwood environment and added the from_django_model to the default sources.
We only need to make sure the logic when either browser keys or config key do not exist is consistent among sources and we are ready to merge

**Review by felipemontoya — APPROVED** [review]

**Review by felipemontoya — APPROVED** [review]
This is ready. Good work

**Pull request merged**